src/SFML/Window/DRM/DRMContext.cpp

```diff
@@ -130,6 +128,306 @@ namespace
+    {
+        for (int i = 0; i < connector.count_encoders; ++i)
+        {
+            const sf::Uint32 encoderId = connector.encoders[i];
+            const drmModeEncoderPtr encoder = drmModeGetEncoder(drm.fd, encoderId);
+
+            if (encoder)
+            {
+                const sf::Uint32 crtcId = findCrtcForEncoder(resources, *encoder);
+
+                drmModeFreeEncoder(encoder);
+                if (crtcId != 0)
+                {
+                    return crtcId;
+                }
+            }
+        }
+
+        // No match found
+        return 0;
+    }
+
+    int getResources(int fd, drmModeResPtr& resources)
+    {
+        resources = drmModeGetResources(fd);
+        if (resources == NULL)
+            return -1;
+        return 0;
+    }
+
+    int hasMonitorConnected(int fd, drmModeRes& resources)
+    {
+        drmModeConnectorPtr connector;
+        for (int i = 0; i < resources.count_connectors; ++i)
+        {
+            connector = drmModeGetConnector(fd, resources.connectors[i]);
+            if (connector->connection == DRM_MODE_CONNECTED)
+            {
+                // There is a monitor connected
+                drmModeFreeConnector(connector);
+                connector = NULL;
+                return 1;
+            }
+            drmModeFreeConnector(connector);
+            connector = NULL;
+        }
+        return 0;
+    }
+
+    int findDrmDevice(drmModeResPtr& resources)
+    {
+        static const int maxDrmDevices = 64;
+
+        drmDevicePtr devices[maxDrmDevices] = { NULL };
+
+        const int numDevices = drmGetDevices2(0, devices, maxDrmDevices);
+        if (numDevices < 0)
+        {
+            sf::err() << "drmGetDevices2 failed: " << std::strerror(-numDevices) << std::endl;
+            return -1;
+        }
+
+        int fd = -1;
+        for (int i = 0; i < numDevices; ++i)
+        {
+            drmDevicePtr device = devices[i];
+            int ret;
+
+            if (!(device->available_nodes & (1 << DRM_NODE_PRIMARY)))
+                continue;
+            // OK, it's a primary device. If we can get the drmModeResources, it means it's also a KMS-capable device.
+            fd = open(device->nodes[DRM_NODE_PRIMARY], O_RDWR);
+            if (fd < 0)
+                continue;
+            ret = getResources(fd, resources);
+            if(std::getenv("SFML_DRM_DEBUG"))
+                sf::err() << "DRM device used: " << i << std::endl;
+
+            if(!ret && hasMonitorConnected(fd, *resources) != 0)
+                break;
+            close(fd);
+            fd = -1;
+        }
+        drmFreeDevices(devices, numDevices);
+
+        if (fd < 0)
+            sf::err() << "No drm device found!" << std::endl;
+        return fd;
+    }
+
+    int initDrm(sf::priv::Drm& drm, const char* device, const char* modeStr, unsigned int vrefresh)
```

Comment:
I want to change `device` and `modeStr` to `std::string` but I'm not sure how to handle line 359.
```cpp
if (modeStr && *modeStr)
```
I suppose this is saying "if the string is not null and the first character is not null"? Is that the right way to read this? Would this translate to `if (!modeStr.empty()`?

Reply:
This `char*` return value is seemingly behaving like a `std::optional<std::string>`. There may be no string, or there may be an empty string. Those are distinct shapes. Because of that, I don't think we can replace this with merely a `std::string` since that would fail to model the distinction between a null string and an empty string.